bfilter.1
.TH BFILTER "1" "June 2016" "bfilter"

.SH NAME
bfilter \- multinomial Bayesian classifier for emails

.SH SYNOPSIS
\fBbfilter\fP [\fB-b\fP] \fBtrain\fP \fIclass\fP < \fIknown-email\fP

\fBbfilter\fP [\fB-b\fP] ( \fBclassify\fP | \fBannotate\fP ) < \fIunknown-email\fP

\fBbfilter\fP [\fB-b\fP] \fBuntrain\fP \fIclass\fP < \fIpreviously-trained-email\fP

\fBbfilter\fP [\fB-b\fP] \fBretrain\fP \fIold-class\fP \fInew-class\fP < \fIpreviously-trained-email\fP

\fBbfilter move\fP \fIold-class-name\fP \fInew-class-name\fP

.SH DESCRIPTION
\fBBfilter\fP is a document classifier for email messages. First it must
be trained on sample messages from different classes. It records
characteristics of the trained messages in a database. Then it can
evaluate new messages and decide which class they are most likely to
belong to.
.PP
One typical application of \fBbfilter\fP is to classify messages as
\fIspam\fP or \fIreal\fP. However, \fBbfilter\fP has no inherent
knowledge of spam characteristics. It could equally well be used to sort
email into, for example, \fIbusiness\fP and \fIpersonal\fP classes,
provided those were the classes used in training.
.PP
Nor is \fBbfilter\fP limited to a two-way classification. A
typical application will use \fBbfilter\fP both to identify spam
messages, and to sort the real messages, for example by training
messages from three classes: \fIbusiness\fP, \fIpersonal\fP, and
\fIspam\fP.
.SS Modes
.PP
The
.B bfilter
command has several operation modes. The first command line argument (apart
from options) specifies the mode. The mode may be abbreviated to any
unambiguous prefix, thus you can write
.B bfilter classify
or
.B bfilter class
or just
.BR "bfilter c" ,
all have the same meaning.
.PP
Most operation modes read a single email message from standard input. If the
\fB-b\fP flag is given, they instead read multiple messages in Berkeley
mailspool format.
.TP
\fBtrain\fP \fIclass\fP
In \fBtrain\fP mode, you are asserting that the input email is an instance of
the given \fIclass\fP.
.B Bfilter
updates its database accordingly. There is no output.
.TP
.B classify
In this mode,
.B bfilter
reads a single email from standard input, and evaluates which class it
is most likely to belong to. It writes two lines of output: the first is
the name of the deduced class; the second is an integer representing
.BR bfilter 's
confidence in the classification. For example, the complete output might
be:
.IP
\f(CWspam\fP
.br
\f(CW17\fP
.IP
See
.I Confidence Estimates
below for more details on the meaning of the number.
.TP
.B annotate
The
.B annotate
mode is similar to
.B classify
mode, except that \fBbfilter\fP copies the email to standard output,
adding an extra header line at the end of the header. The new header
line contains the deduced class, and the confidence estimate in
parentheses. For example:
.IP
\f(CWX-Bfilter-Class: spam (17)\fP
.TP
\fBuntrain\fP \fIclass\fP
The \fBuntrain\fP mode undoes the effect of a previous \fBtrain\fP
command. The intended use of \fBuntrain\fP \fIclass\fP is for the case
where you have previously trained a message as belonging to \fIclass\fP,
and have since changed your mind and decided that it does not in fact
belong to \fIclass\fP. See \fIUntraining\fP below for more information. 
.TP
\fBretrain\fP \fIold-class new-class\fP
This is equivalent to \fBuntrain\fP \fIold-class\fP followed by \fBtrain
\fInew-class\fP.
.TP
\fBmove\fP \fIold-name new-name\fP
The \fBmove\fP mode can be used to rename an existing class.

.SH FILES
The default database is \fB.bfilter.sqlite\fP in \fB$HOME\fP. An alternate
database can be specified in the \fBBFILTER_DB\fP environment variable.

.SH NOTES
.SS Confidence Estimates
When \fBbfilter\fP is classifying a message\(emthat is in \fBclassify\fP
and \fBannotate\fP modes\(emit outputs both the classification and a
confidence estimate. The confidence estimate is a non-negative integer:
the larger it is, the more confident \fBbfilter\fP is of the
classification. The lowest possible confidence estimate is \fB0\fP
indicating that \fBbfilter\fP's classification is highly uncertain. (As
a special case, if \fBbfilter\fP's database contains no messages, it
outputs a class of \fBUNKNOWN\fP and a confidence estimate of \fB0\fP.)
.PP
The confidence is scaled to represent an \fIapproximate\fP percentage,
so \fB100\fP represents an extremely confident classification. However,
in the current implementation, the number reported can exceed 100. The
exact mathematical meaning of the confidence estimate is somewhat murky,
and this is not the place to delve into it. The way it is calculated
may well change in the future, but it is intended that it will remain an
integer, with \fB0\fP meaning very uncertain, and \fB100\fP very
certain.
.PP
It is strongly suggested that the confidence estimate is used to drive
user interaction. If \fBbfilter\fP classifies a message as \fBspam\fP
with a low confidence\(emsay, less than \fB20\fP\(emit would be wise to
present the message to the user and ask them to check the
classification. Then, whether the user agrees or disagrees with
\fBbfilter\fP, train the message according to their input.
.SS Untraining
In \fBuntrain\fP and \fBretrain\fP modes, \fBbfilter\fP will ``unlearn''
the characteristics of the input message. This is to allow
classification mistakes to be corrected.
.PP
Currently, \fBbfilter\fP must be told which class it is untraining the
message from, and it makes no attempt to ensure that the message was
previously trained to that class.  Typically, \fBbfilter\fP will be a
component in a system which records whether a message has been trained,
and to which class, and can thus ensure that \fBuntrain\fP mode is
called appopriately.
.PP
Calling \fBuntrain\fP inappropriately should not be catastrophic. For
example, it should not cause \fBbfilter\fP to crash, and it should only
affect its judgements in reasonably sensible ways. However, this has not
been extensively tested, and it is probably wise not to untrain messages
unless you are reasonably sure that they were indeed previously trained
to the stated class.
.PP
Future versions of \fBbfilter\fP are likely to be more robust in this
area.

.SH SEE ALSO
.BR sqlite3 (1)
.PP
https://github.com/TobyGoodwin/bfilter
.br
http://www.paulgraham.com/antispam.html
.br
http://nlp.stanford.edu/IR-book/html/htmledition/naive-bayes-text-classification-1.html
.br
http://mpv6.com/posts/2015-11-15-bfilter.html

.SH COPYRIGHT
Copyright \(co 2003, 2004 Chris Lightfoot.
.br
Copyright \(co 2015 - 2016 Toby Goodwin.
.br
toby@paccrat.org
.br
https://github.com/TobyGoodwin/bfilter
.PP
Bfilter is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your
option) any later version.
.PP
Bfilter is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
for more details.
.PP
You should have received a copy of the GNU General Public License along
with bfilter.  If not, see <http://www.gnu.org/licenses/>.


CHANGES
Change log for bfilter
$Id: CHANGES,v 1.9 2006/05/20 16:10:44 chris Exp $

Fixed a crash bug in cleandb. Record log of `spam probaility' in
X-Spam-Probability header.

0.4

Fixed typos in man page. The help message now prints the bfilter version. Added
token history code, so that `Buy Cheap Viagra' now gives rise to the tokens
`buy%cheap', `cheap%viagra' and `buy%cheap%viagra', in addition to the
single-word ones used previously. Added a stats option for printing information
about the contents of the database. Changed cleandb to work by copying the
retained items to a new database file, then renaming it over the old one; this
is faster than the old method and should ensure that disk space is reclaimed
properly.

0.3

Fixed cleandb code. More error indications. Added tokeniser state machine
diagram, mostly as an experiment in graphviz. Uses TDB rather than GDBM, for
better concurrency. Does not use _ as a token separator when parsing
X-Spam-Status: headers, so bfilter can make better use of test results added
by Spam Assassin.

0.2

Added timestamps to database keys. Added cleandb option. Added X-Spam-Status:
(from Spam Assassin) to the list of headers to inspect. Changed the base
all-spam / all-real probabilities to 0.00001 / 0.99999.

0.1

Initial public release.


bfilter.c
/*
 * bfilter.c:
 * Simple Bayesian email filter, in C.
 *
 * Copyright (c) 2003 Chris Lightfoot. All rights reserved.
 * Email: chris@ex-parrot.com; WWW: http://www.ex-parrot.com/~chris/
 *
 */

static const char rcsid[] = "$Id: bfilter.c,v 1.24 2005/06/07 16:41:22 chris Exp $";

#include <sys/types.h>

#include <errno.h>
#include <math.h>
#include <pwd.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <time.h>
#include <unistd.h>

#include <netinet/in.h>     /* for ntohl/htonl */

#include "bfilter.h"
#include "compose.h"
#include "db.h"
#include "skiplist.h"
#include "token.h"
#include "util.h"

/* unbase64 CHAR
 * Decode a single base64 CHAR. */
uint32_t unbase64(char c) {
    static const char *s = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    if (c == '=')
        return 0;
    else
        return (uint32_t)(strchr(s, c) - s);
}

/* decode_base64 BUFFER LENGTH
 * Decode LENGTH characters of base64 data starting at BUFFER in place,
 * returning the number of bytes of real data produced, and padding the
 * end of the original extent with whitespace. */
size_t decode_base64(char *buf, size_t len) {
    char *rd, *wr;

    for (rd = buf, wr = buf; rd < buf + len; rd += 4, wr += 3) {
        uint32_t X;
        X = unbase64(rd[3]) | (unbase64(rd[2]) << 6) | (unbase64(rd[1]) << 12) | (unbase64(rd[0]) << 18);
        wr[2] = X & 0xff;
        wr[1] = (X >> 8) & 0xff;
        wr[0] = (X >> 16) & 0xff;
    }

    memset(wr, ' ', len - (wr - buf));

    return wr - buf;
}

/* is_b64_chars BUF LEN
 * Are the first LEN bytes of BUF composed only of base64 characters? */
int is_b64_chars(const char *buf, size_t len) {
    const char *p;
    while (len > 0 && buf[len - 1] == '=')
        --len;
    for (p = buf; p < buf + len; ++p)
        if (!((*p >= 'A' && *p <= 'Z')
               || (*p >= 'a' && *p <= 'z')
               || (*p >= '0' && *p <= '9')
               || *p == '+' || *p == '/'))
            return 0;
    return 1;
}

/* read_email FROMLINE PASSTHROUGH STREAM TEMPFILE
 * Read an email from STREAM, tokenize and submit.
 * Stops at end of file, or, if FROMLINE is nonzero, when a "\n\nFrom " line
 * is found. If PASSTHROUGH is nonzero, we emit the headers to standard output
 * and save the body of the email in a temporary file, which we return as a
 * stdio stream in *TEMPFILE; in this case, any X-Spam-Probability: header is
 * discarded from the email. Returns 1 on success or 0 on failure. */
size_t nbytesrd;
int read_email(const int fromline, const int passthrough, FILE *fp, FILE **tempfp) {
    static size_t buflen;
    static char *buf;
    int i, j;
    enum parsestate { hdr = 0, hdr_xsp, hdr_rel, bdy_blank, bdy, bdy_b64_1, bdy_b64, end } state = hdr;
    static char *b64buf = NULL;
    static size_t b64alloc;
    size_t b64len = 0, b64linelen = 0;

    compose_reset();

    /* 
     * Various tests we use to drive the state machine.
     */
#   define is_blank()      (j == 0)
#   define starts_nwsp()   (j > 0 && !strchr(" \t", buf[0]))
#   define is_hdr_xsp()    ((j >= 19 && strncasecmp(buf, "X-Spam-Probability:", 19) == 0)   \
                            || (j >= 13 && strncasecmp(buf, "X-Spam-Words:", 13) == 0))
#   define is_hdr_rel()    ((j >= 5 && strncasecmp(buf, "From:", 5) == 0)           \
                            || (j >= 8 && strncasecmp(buf, "Subject:", 8) == 0)     \
                            || (j >= 3 && strncasecmp(buf, "To:", 3) == 0)          \
                            || (j >= 14 && strncasecmp(buf, "X-Spam-Status:", 14) == 0))
#   define is_from_()      (fromline && j >= 5 && strncmp(buf, "From ", 5) == 0)
#   define is_b64()        is_b64_chars(buf, j)

    /*
     * Various things to do with identifying and parsing base64 data.
     */

    /* Reset the base64 buffer. */
#   define b64_reset()  \
            do {                    \
                b64linelen = 0;     \
                b64len = 0;         \
            } while (0)
    
    /* Save some candidate base64 data. */
#   define b64_save()    \
            do {                                                                \
                if (!b64buf || b64len + j > b64alloc)                           \
                    b64buf = xrealloc(b64buf, b64alloc = 2 * (b64len + j));     \
                memcpy(b64buf + b64len, buf, j);                                \
                b64len += j;                                                    \
            } while (0)

    /* Decode and submit data from the base64 buffer. */
#   define b64_submit()   \
            do {                                            \
                size_t n, m;                                \
                n = b64len & ~3;                            \
                m = decode_base64(b64buf, n);               \
                tokenize(b64buf, m, 0);       \
                memmove(b64buf, b64buf + n, b64len - n);    \
                b64len -= n;                                \
            } while (0)

    /* Submit the data from the base64 buffer undecoded, and reset the
     * buffer. */
#   define b64_submit_un()                                 \
            do {                                           \
                tokenize(b64buf, b64len, 0); \
                b64_reset();                               \
            } while (0)
                            

    if (passthrough && !(*tempfp = tmpfile()))
        return 0;

    if (!buf)
        buf = xmalloc(buflen = 1024);
    
    do {
        /* Obtain a line from the email. */
        j = 0;
        while ((i = getc(fp)) != EOF) {
            buf[j++] = (char)i;
            ++nbytesrd;
            if (j == buflen)
                buf = xrealloc(buf, buflen *= 2);
            if (i == '\n') break;
        }

        --j; /* j is now the number of non-\n characters */
        
        if (ferror(fp))
            goto abort;

        /*
         * State machine. This is quite complicated. See tokeniser-states.dot
         * for a pretty (albeit not-very-helpful) diagram.
         */
        switch (state) {
            /* hdr: general header */
            case hdr:
                if (is_blank())
                    state = bdy_blank;
                else if (is_hdr_rel())
                    state = hdr_rel;
                else if (is_hdr_xsp())
                    state = hdr_xsp;
                break;

            /* hdr_rel/hdr_xsp: special header */
            case hdr_rel:
            case hdr_xsp:
                if (is_blank())
                    state = bdy_blank;
                else if (is_hdr_rel())
                    state = hdr_rel;
                else if (is_hdr_xsp())
                    state = hdr_xsp;
                else if (starts_nwsp())
                    state = hdr;
                break;

            /* bdy_blank: blank line in body */
            case bdy_blank:
                if (is_from_())
                    state = end;
                else if (j > 32 && is_b64()) {
                    b64_reset();
                    b64_save();
                    b64linelen = j;
                    state = bdy_b64_1;
                } else if (!is_blank())
                    state = bdy;
                break;

            /* bdy: nonblank nonbase64 line in body */
            case bdy:
                if (is_blank())
                    state = bdy_blank;
                break;

            /* bdy_b64_1: candidate first line of base64 data */
            case bdy_b64_1:
                if (is_blank()) {
                    b64_submit_un();
                    state = bdy_blank;
                } else if (j > b64linelen || !is_b64()) {
                    b64_submit_un();
                    state = bdy;
                } else {
                    b64_save();
                    b64_submit();
                    state = bdy_b64;
                }
                break;

            /* bdy_b64: line inside base64 data */
            case bdy_b64:
                if (is_blank())
                    state = bdy_blank;
                else if (j > b64linelen || !is_b64()) {
                    b64_submit_un();
                    state = bdy;
                } else {
                    b64_save();
                    b64_submit();
                }
                break;

            case end:
                /* No way out. */
                break;
        }

        /* End of file ends the email. */
        if (feof(fp)) {
            if (state != bdy && state != bdy_blank)
                goto abort;
            else
                state = end;
        }

        /* Possibly emit/save text. */
        if (passthrough) {
            switch (state) {
                case hdr:
                case hdr_rel:
                    if (fwrite(buf, 1, j + 1, stdout) != j + 1)
                        goto abort;
                    break;

                case hdr_xsp:
                case end:
                    break;

                default:
                    if (fwrite(buf, 1, j + 1, *tempfp) != j + 1)
                        goto abort;
                    break;
            }
        }
        
        /* Pass stuff to the tokeniser. */
        if (state == hdr_rel || state == bdy) {
            char *p;
            size_t len;
            p = NULL;
            len = (size_t)j;
            if (state == hdr_rel && starts_nwsp() && (p = strchr(buf, ':'))) {
                ++p;
                len -= p - buf;
            } if (!p)
                p = buf;

            tokenize(p, len, state == hdr_rel);
        }
    } while (state != end);
    
    return 1;
    
abort:
    if (passthrough && *tempfp)
        fclose(*tempfp);
    return 0;
    
}

